bin/test/main.ml

```diff
@@ -306,73 +364,41 @@ let run_exn (`Setup ()) (`Non_deterministic non_deterministic)
-                 (* Skip raw blocks. *)
-                 | true, _, _, Raw -> Block.pp ?syntax ppf t
-                 (* The command is not active, skip it. *)
-                 | false, _, _, _ -> Block.pp ?syntax ppf t
-                 (* the command is active but non-deterministic so skip everything *)
-                 | true, false, `Non_det `Command, _ -> Block.pp ?syntax ppf t
-                 (* the command is active but it's output is
-                    non-deterministic; run it but keep the old output. *)
-                 | true, false, `Non_det `Output, Cram { tests; _ } ->
-                   Block.pp ?syntax ppf t;
-                   let blacklist = Block.unset_variables t in
-                   List.iter (fun t ->
-                       let _: int = run_test ?root blacklist temp_file t in ()
-                     ) tests
-                 | true, false, `Non_det `Output, Toplevel tests ->
-                   assert (syntax <> Some Cram);
-                   Block.pp ppf t;
-                   List.iter (fun test ->
-                       match eval_test t ?root c test with
-                       | Ok _    -> ()
-                       | Error e ->
-                         let output = List.map output_from_line e in
-                         if Output.equal test.output output then ()
-                         else err_eval ~cmd:test.command e
-                     ) tests
-                 (* Run raw OCaml code *)
-                 | true, _, _, OCaml ->
-                   assert (syntax <> Some Cram);
-                   (match Block.file t with
-                    | Some ml_file ->
-                      update_file_or_block ?root ppf file ml_file t direction
-                    | None ->
-                      eval_raw t ?root c ~line:t.line t.contents;
-                      Block.pp ppf t )
-                 (* Cram tests. *)
-                 | true, _, _, Cram { tests; pad } ->
-                   run_cram_tests ?syntax t ?root ppf temp_file pad tests
-                 (* Top-level tests. *)
-                 | true, _, _, Toplevel tests ->
-                   assert (syntax <> Some Cram);
-                   match Block.file t with
-                   | Some ml_file ->
-                     update_file_or_block ?root ppf file ml_file t direction
-                   | None ->
-                     run_toplevel_tests ?root c ppf tests t
-              )
+                (fun () -> test_block ~ppf ~temp_file t)
+            with Failure msg ->
+              raise (Test_block_failure (t, msg))
         ) items;
       Format.pp_print_flush ppf ();
       Buffer.contents buf);
   Hashtbl.iter (write_parts ~force_output) files;
   0
 
+let report_error_in_block block msg =
+  let kind =
+    match block.Block.value with
+    | Raw | Error _ -> ""
+    | OCaml -> "OCaml "
+    | Cram _ -> "cram "
+    | Toplevel _ -> "toplevel "
+  in
+  Fmt.epr "Error in the %sblock in %s at line %d:@]\n%s\n"
```

Comment:
Could you make that "%scode block" ? I think it's the right terminology for those, block is too generic!

Reply:
This will be "OCaml block" or "toplevel block" already. This may be too much words. Also "block" is not generic at all.

Reply:
Agreed "code block" is better.